### PR TITLE
optimization for HID class

### PR DESCRIPTION
**Pull request by jiadingxu**
Hello hathach:
     Here're some optimizatoins for the HID class and a sample configuraiton.

### CODE REVIEW

File: src/class/hid/hid_device.h [hunk 1]
```diff
@@ -380,7 +380,30 @@ static inline bool  tud_hid_gamepad_report(uint8_t report_id, int8_t x, int8_t y
 // HID Generic Input & Output
 // - 1st parameter is report size (mandatory)
 // - 2nd parameter is report id HID_REPORT_ID(n) (optional)
-#define TUD_HID_REPORT_DESC_GENERIC_INOUT(report_size, ...) \
+#if (CFG_TUSB_REPORT_ID_COUNT == 2)
+  #define TUD_HID_REPORT_DESC_GENERIC_INOUT(report_size, ...) \
```
Comment (hathach) [inline src/class/hid/hid_device.h:384]: please explain what is this template used for ?

File: src/device/dcd.h [hunk 2]
```diff
@@ -82,6 +83,11 @@ typedef struct TU_ATTR_ALIGNED(4)
       uint32_t frame_count;
     }sof;
 
+    // Plugged
+    struct {
```
Comment (hathach) [inline src/device/dcd.h:87]: please explain why adding this change

File: examples/device/hid_generic_inout/src/tusb_config.h [hunk 3]
```diff
@@ -97,6 +97,26 @@
...
+#if REPORT_ID_COUNT == 2
+    #define REPORT_ID_OUT           1
+    #define REPORT_ID_IN            2
+#elif REPORT_ID_COUNT == 1
+  #define REPORT_ID               1
+  #define REPORT_ID_OUT           REPORT_ID
+  #define REPORT_ID_IN            REPORT_ID
+#endif
+
+#ifndef CFG_TUSB_REPORT_ID_COUNT
```
Comment (hathach) [inline examples/device/hid_generic_inout/src/tusb_config.h:116]: report ID count is application-specific and should not be part of configuration parameter.

**Review by hathach — COMMENTED** [review]
thank you for your PR, please check out review for my questions on the changes.

File: src/class/hid/hid_device.h (hunk 1 above)
Comment (jiadingxu) [reply src/class/hid/hid_device.h:384]: I found that the macro didn't support two different Report IDs(in / out), so I tried to separate them.

Comment (jiadingxu) [reply src/class/hid/hid_device.h:384]: For example: If I need a report id(1) for out and another one(2) for in,  I won't get the data throuth this original macro.

Comment (jiadingxu) [reply src/class/hid/hid_device.h:384]: During designing a usb generic HID communication, I found that the macro didn't support two different Report IDs(in / out), so I tried to separate them.
For example: 
If I need a report id(Report ID: 1) for out reports and another one(Rreport ID: 2) for in reports, I won't get the data through this original macro.

Comment (hathach) [reply src/class/hid/hid_device.h:384]: if template descriptor does not fit your need, you don't need to use it, just make your own descritpor. These template only cover the most common usage by user. Therefore this should be reverted !

**Comment by HiFiPhile** [comment]
Close deperated PRs.

**Pull request closed without merging**